### ISSUE TITLE
Switch tree navigation from breadth-first to depth-first

### ISSUE THREAD

**Issue by granawkins** [kstaats/karoo_gp#89]
Karoo Classic uses a breadth-first-search ordering (`BFS`) when navigating trees. The new `Node` API makes heavy use of recursion, which is by nature depth-first (`DFS`). We've implemented workarounds for breadth-first to guarantee continuity of test results through all the recent changes.

We've decided to switch to DFS by default, and remove support for BFS because:
- We'd prefer to choose just one to support, so we don't have to write/update everything twice.
- BFS is theoretically faster and light, requires less code, and we use a lot of recursion already.

NOTE: We've discussed that Python handles recursion a bit differently, so there are some implicit limits. However, we don't expect to reach those limits (works fine at `tree_depth_base=15`), and if we do, they can be increased manually.

### What's the impact on Trees?
'Grow' trees are generated stochastically. Starting from the root node, at every 'child' position, flip a coin to decide whether to add a function (with additional children) or a terminal (no children). So they result in irregularly-shaped trees.

We'd expect that a population of grow trees generated with `BFS` and `DFS` will have different 'shapes'. Here's a diagram showing that `DFS` trees have a higher probability of being *deep* and *wide* (i.e. have lots of children).
![dfs_bfs_diagram](https://user-images.githubusercontent.com/50287275/186341562-b03b0a87-8217-4f7d-a2f2-5410a9888e4b.png)

I did an experiment where I generated 1000-each trees using `DFS` and `BFS`, and compare the frequency of depths and widths. Checks out.
![dfs_bfs_chart](https://user-images.githubusercontent.com/50287275/186341576-9f3bf027-c54c-4f1f-8ef4-9a35fc00cf86.png)

So, expect a PR with this soon.

**Comment by ezio-melotti** [comment]
Just out of curiosity, have you tried to reach Python's recursion limit by increasing the tree depth? Also have you noticed any performance differences between the two algos on trees with a similar number of nodes?

**Comment by kstaats** [comment]
Very, very interesting. I am concerned for the lack of breadth in the 
new DFS method. Let's discuss how this might affect solutions.

Thank you!

**Comment by granawkins** [comment]
I found some mistakes in the `DFS` algorithm - the results above weren't actually correct, and didn't match what I was saying. These results do:
![dfs_bfs_chart_2](https://user-images.githubusercontent.com/50287275/186350991-2cca45c1-9886-4bd9-acda-09e4fb3d3872.png)

So as expected, DFS are on average a tiny bit *deeper* and *more children*, but much less than the previous chart. Let's discuss anyway.

@ezio-melotti When I was testing recursion, I got bored waiting around depth 17/18, so I never reached the limit. I will try some performance tests now.
